### PR TITLE
Implements function signatures

### LINKED ISSUE #86: Complete optional parameters for functions

### PR DESCRIPTION
**Pull request by bthesorceror**
- User defined
- Imported
- Builtins

Closes #86

### CODE REVIEW

File: src/protocol/requests.rs [hunk 1]
```diff
@@ -138,7 +139,30 @@ pub struct CompletionParams {
...
-    trigger_character: Option<String>,
+    pub trigger_character: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+pub struct SignatureHelpParams {
+    #[serde(rename = "textDocument")]
+    pub text_document: TextDocumentIdentifier,
+    pub position: Position,
+    pub context: Option<SignatureHelpContext>,
```
Comment (kelwang) [inline src/protocol/requests.rs]: there is an additional parameters in the specs
```
        /**
	 * An optional token that a server can use to report work done progress.
	 */
	workDoneToken?: ProgressToken;
```

File: src/protocol/requests.rs [hunk 2]
```diff
@@ -138,7 +139,30 @@ pub struct CompletionParams {
...
+pub enum SignatureHelpTriggerKind {
+    Invoked = 1,
+    TriggerCharacter = 2,
+    ContentChange = 3,
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+pub struct SignatureHelpContext {
+    pub is_retrigger: bool,
+    pub trigger_character: Option<String>,
```
Comment (kelwang) [inline src/protocol/requests.rs]: should this be a char? when serialized, it will be a string in json

Comment (bthesorceror) [reply src/protocol/requests.rs]: There is no difference in JSON between a char and a string.

Comment (kelwang) [reply src/protocol/requests.rs]: cool, so char should be better here

File: src/protocol/requests.rs (hunk 1 above)
Comment (bthesorceror) [reply src/protocol/requests.rs]: LSP clients don't seem to be sending this and since its optional I opted to omit it for the time.

**Pull request merged**